public/pages/CreateIndex/components/IndexMapping/IndexMapping.tsx

```diff
@@ -0,0 +1,366 @@
+          },
+          {
+            label: readonly ? "JSON" : "JSON Editor",
+            id: EDITOR_MODE.JSON,
+            "data-test-subj": "editorTypeJsonEditor",
+          },
+        ]}
+      />
+      <EuiSpacer />
+      {editorMode === EDITOR_MODE.VISUAL ? (
+        <>
+          {transformedTreeItems.length ? (
+            <EuiTreeView
+              key={renderKey}
+              expandByDefault={!readonly}
+              className="index-mapping-tree"
+              aria-labelledby="label"
+              items={transformValueToTreeItems(value)}
+            />
+          ) : (
+            <p>You have no field mappings.</p>
+          )}
+          {readonly ? null : (
+            <>
+              <EuiSpacer />
+              <EuiButton style={{ marginRight: 8 }} data-test-subj="createIndexAddFieldButton" onClick={() => addField("")}>
+                Add new field
+              </EuiButton>
+              <EuiButton
+                data-test-subj="createIndexAddObjectFieldButton"
+                onClick={() =>
+                  addField("", {
+                    type: "object",
+                  })
+                }
+              >
+                Add new object
+              </EuiButton>
+            </>
+          )}
+        </>
+      ) : (
+        <>
+          {isEdit && !readonly && !isEmpty(oldValue) ? (
+            <>
+              <EuiButton
+                size="s"
+                data-test-subj="previousMappingsJsonButton"
+                onClick={() => {
+                  Modal.show({
+                    title: "Previous mappings",
+                    content: (
+                      <JSONEditor
+                        readOnly
+                        value={JSON.stringify(
+                          {
+                            ...oldValue,
+                            properties: transformArrayToObject(oldValue?.properties || []),
+                          },
+                          null,
+                          2
+                        )}
+                      />
+                    ),
+                    "data-test-subj": "previousMappingsJsonModal",
+                    onOk: () => {},
+                  });
+                }}
+              >
+                See previous settings
+              </EuiButton>
+              <EuiSpacer />
+            </>
+          ) : null}
+          {readonly ? (
+            <JSONEditor
+              ref={JSONEditorRef}
+              value={JSON.stringify(
+                {
+                  ...propsValue,
+                  properties: transformArrayToObject(value || []),
+                },
+                null,
+                2
+              )}
+              readOnly={readonly}
+              width="100%"
+            />
+          ) : (
+            <CustomFormRow
+              data-test-subj="mappingsJsonEditorFormRow"
+              label="Specify index mapping"
+              helpText={
+                <div>
```

Comment:
Should we use a fragment instead?

Reply:
Agree, fixed on https://github.com/opensearch-project/index-management-dashboards-plugin/commit/9c23ee35505cdf9d771d3f28c949576dc0e1baeb.